### DISCUSSION TITLE
Language choice discussion

### DISCUSSION THREAD

**Discussion by lambdanil** [ashos/astos#89] [Ideas]
Hi everyone,

With AshOS further maturing I think it's a good time to reconsider our language choice.

When the project originally started I chose Python as the implementation language - simply because it's good for quick prototypes and has a lot of features built-in. But now the scope of the project is much bigger than first expected, so it's a good time to ask if Python is the right language for this project.

One of the disadvantages of Python is it's reliance on a language runtime - if the Python installation is somehow messed up, the entire system falls apart. For that reason a compiled language might be better suited.

Here's a small list of the pros and cons of different approaches:

#### Cons
* Means an entire rewrite, obviously very big changes
* We would have to distribute the precompiled binaries (we could however just host them as releases on the github repo)
#### Pros
* More reliable - not reliant on a language runtime = fewer points of failure
* Harder for the user to break
* Less dependencies - smaller base installation
* We could achieve better interoperability with the system, using native system calls, native library for interacting with btrfs, etc.

#### Cons
* Shell scripts aren't very well suited for bigger projects
#### Pros
* It's more unix-y
* More modular

#### Cons
* Reliant on a language runtime
* More dependencies
#### Pros
* Much less effort than moving to an entirely different language
* Python is quick and easy to write, which makes it easy to contribute and add new features

---

I'd like to discuss the merits of these approaches, and choose whichever one works best for the project. Personally I think approaches 1 and 3 are the more interesting ones.

I'll be waiting for any input from our community!

**Comment by i2** [comment]
In this discussion, I would make a distinction between `AshOS installer` and Ash itself (refer to [this discussion](https://github.com/ashos/ashos/discussions/63)).

Installer can stay in Python as performance isn't important and python give us pluggability and ease of development.
Just to clarify for participants, this question is more relevant to Ash itself.

**For Ash itself, I'd like to envision AshOS in C, and that's proper C not OOP C++.**

**Reply by xplshn** [reply]
I would suggest Go, because it is (like)C but convenient. And it doesn't have the dependency nightmare of C, and it can be easily compiled as a static binary too, and there is also the fact that one can compile it using LLVM/GCC if using CGO, also Go is very readable, unlike C++.

**Comment by i2** [comment]
@ochibani In your experience, what has been the pain point(s) with converting ashos to Rust so far?

**Reply by ochibani** [reply]
It was replacing 'python-anytree' and I didn't fix this, as the code still depends on this Python library.😅

**Reply by i2** [reply]
I can see [this](https://doc.rust-lang.org/nightly/rustc/platform-support.html), but one aspect that I'd like to be considered: Does rust run on **all** the **OSes** and **architectures** that Python supports?

**Reply by ochibani** [reply]
Why not?

**Reply by xplshn** [reply]
Another cons against Rust would be that it is quite contrary to this:
> Python is quick and easy to write, which makes it easy to contribute and add new features

**Comment by xplshn** [comment]
> We would have to distribute the precompiled binaries (we could however just host them as releases on the github repo)

Is there a serious problem with this? Statically compiling Rust is not a big problem, however, Rust binaries that are compiled this way are usually bigger than other language's because they bundle everything, not just the necessary symbols. C, C++, Go, Zig, etc, simply bundle the symbols needed, not everything.

**Comment by ochibani** [comment]
# Changes in Rust-Related PR

Hi everyone,

I’d like to discuss the changes introduced in this [Rust-related PR](https://github.com/ashos/ashos/pull/133).

## Key Changes:

1. **Splitting AshOS into components:**
   - **Ash Installer** (experimental)
   - **Ash Commands**
   - **Ash Profiles**
   - **Ash ISO** (planned for the future)

2. **Management of Ash Binary:**
   - Installation and updates are now handled by the system's package manager, with Ash itself focusing solely on snapshot management.

3. **Snapshot Enhancements:**
   - Snapshots now include the `/var` directory, alongside `/etc` and `/boot`.

4. **Profile Management:**
   - Added `/etc/ash/profile`, allowing users to manage installed packages and services.

5. **Rebuild Command:**
   - Introduced the `rebuild` command, which lets users bootstrap an existing snapshot. This is particularly useful for outdated snapshots that are difficult to upgrade.

6. **Import & Export Commands:**
   - Added commands to import and export snapshots between systems.

7. **Reset Command:**
   - Introduced a `reset` command, which functions as a factory reset for the system.

8. **Recovery Mode:**
   - Added a recovery mode option to the GRUB menu.

9. **Multi-Distribution Snapshots:**
   - Ash now supports snapshots with different distributions. For example, you can have Arch Linux in one snapshot (snapshot-1) and Debian in another (snapshot-2). You can even switch distributions (e.g., from Debian to Arch) without reformatting the hard drive.

10. **Decoupling Core Components:**
    - The bootloader, file system, init system, and package manager have been separated from the main codebase.

11. **Additional Improvements:**
    - Various minor changes and enhancements to existing commands.

## Final Thoughts:

If there were a programming language with the simplicity of Bash or Python but better error handling, it might be a better fit for this project. Honestly, I'm not entirely happy with my Rust code—it feels more like a Bash script written in Rust 😅.

The good news is that the code is functional and ready for testing.

**Reply by i2** [reply]
Great work @ochibani. I have been following your work since inception and will look closer at the topics you brought up above.

For the time being, one quick thing that comes to my mind regarding "Ash ISO (planned for the future)":
I had been wary of going in this direction as I didn't want our project to be just like those _numerous_ projects that just slap a few themes and customizations together and call it a "distro". Having said that, I am all ears if there are good arguments for this. What are your reasons for this? (to keep things tidy, please reply in this thread)

And regarding bash-infected Rust code, as mentioned in the readme, same issue exists with the python version!

**Reply by ochibani** [reply]
The answer is very simple, I would suggest creating a custom GUI installer (probably Calamares-based).

**Reply by ochibani** [reply]
The [GUI](https://github.com/ochibani/ashos-calamares) installer is ready for testing on any Arch-based ISO.
